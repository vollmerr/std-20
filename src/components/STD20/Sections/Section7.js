import React from 'react';
import { FieldArray } from 'redux-form';
import {
  FieldDate,
  FieldInput,
  FieldRadioButtons,
  Button,
} from 'state-template';

import schema from '../schema';

class Section7 extends React.Component {
  // attach array indexed name to field name (ex: orderDetails[1].quantity)
  withNamePrefix = (detail, values) => ({
    ...values,
    name: `${detail}.${values.name}`,
  })

  renderDetails = ({ fields }) => {
    if (!fields.length) {
      fields.push({});
    }

    return (
      <fieldset className={'form-group'}>
        <div className={'row align-items-center'}>
          <legend className={'col-md-6'}>7. Order Detail</legend>

          <div className={'col-md-6'}>
            <Button
              text={'Add Request'}
              variant={'primary'}
              className={'float-right m-b-md d-print-none'}
              iconProps={{ name: 'plus-line' }}
              onClick={() => fields.push({})}
            />
          </div>

          {
            fields.map((detail, index) => (
              <React.Fragment key={detail}>
                <FieldRadioButtons {...this.withNamePrefix(detail, schema.orderType)} className={'col-md-8'} inline />
                <FieldDate {...this.withNamePrefix(detail, schema.serviceDate)} className={'col-md-4'} />

                <FieldInput {...this.withNamePrefix(detail, schema.quantity)} className={'col-md-4'} />
                <FieldInput {...this.withNamePrefix(detail, schema.recurringCost)} className={'col-md-4'} />
                <FieldInput {...this.withNamePrefix(detail, schema.nonRecurringCost)} className={'col-md-4'} />

                <FieldInput {...this.withNamePrefix(detail, schema.stateContractNumber)} className={'col-md-4'} />
                <FieldInput {...this.withNamePrefix(detail, schema.featureId)} className={'col-md-4'} />
                <FieldInput {...this.withNamePrefix(detail, schema.billingAccount)} className={'col-md-4'} />

                <FieldInput {...this.withNamePrefix(detail, schema.orderDescription)} className={'col-md-12'} />

                <FieldInput {...this.withNamePrefix(detail, schema.orderComment)} className={'col-md-12'} />

                <div className={'col-md-12'}>
                  <Button
                    text={'Remove Request'}
                    variant={'default'}
                    className={'float-right d-print-none'}
                    iconProps={{ name: 'close-line' }}
                    onClick={() => fields.remove(index)}
                  />
                </div>

                {index !== fields.length - 1 && <div className={'col-md-12'}><hr /></div>}
              </React.Fragment>
            ))
          }
        </div>
      </fieldset>
    );
  }

  render() {
    return (
      <FieldArray name={schema.orderDetails.name} component={this.renderDetails} />
    );
  }
}

export default Section7;
