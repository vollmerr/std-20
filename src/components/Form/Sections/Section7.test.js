import React from 'react';
import { shallow } from 'enzyme';

import Section7 from './Section7';

const props = {};

let wrapper;
let instance;
describe('Section7', () => {
  beforeEach(() => {
    wrapper = shallow(<Section7 {...props} />);
    instance = wrapper.instance();
  });

  describe('withNamePrefix', () => {
    it('should prefix the `name` with the detail key', () => {
      const detail = 'detailKey';
      const values = { id: '123', name: 'test-name' };
      const prefixedName = instance.withNamePrefix(detail, values);
      expect(prefixedName).toEqual({ id: '123', name: 'detailKey.test-name' });
    });
  });

  describe('renderDetails', () => {
    it('should render a empty detail if none provided', () => {
      const Details = shallow(instance.renderDetails({ fields: [] }));
      expect(Details.find('[data-testid^="detail-"]').length).toBe(1);
    });

    it('should render the provided details', () => {
      const Details = shallow(instance.renderDetails({ fields: ['detail-1', 'detail-2'] }));
      expect(Details.find('[data-testid^="detail-"]').length).toBe(2);
    });

    it('should render correctly', () => {
      const Details = shallow(instance.renderDetails({ fields: ['detail-1', 'detail-2'] }));
      expect(Details).toMatchSnapshot();
    });
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
