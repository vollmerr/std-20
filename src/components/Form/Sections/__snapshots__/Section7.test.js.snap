// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section7 renderDetails should render correctly 1`] = `
<div
  aria-labelledby="section7__header"
  className="row align-items-center"
  role="group"
>
  <div
    className="col-md-6"
    id="section7__header"
  >
    7. Order Detail
  </div>
  <div
    className="col-md-6"
  >
    <Button
      active={false}
      className="float-right m-b-md d-print-none"
      iconProps={
        Object {
          "name": "plus-line",
        }
      }
      onClick={[Function]}
      tag="button"
      text="Add Request"
      type="button"
      variant="primary"
    />
  </div>
  <div
    aria-labelledby="section7__header--0"
    className="d-flex flex-wrap w-100"
    data-testid="detail-0"
    key="detail-1"
    role="group"
  >
    <div
      className="col-md-12"
      id="section7__header--0"
    >
      Request #1
    </div>
    <Component
      className="col-md-8"
      inline={true}
      label="Order Type"
      name="detail-1.orderType"
      options={
        Array [
          Object {
            "label": "Add",
            "value": "add",
          },
          Object {
            "label": "Disconnect",
            "value": "disconnect",
          },
          Object {
            "label": "Change",
            "value": "change",
          },
          Object {
            "label": "Move",
            "value": "move",
          },
        ]
      }
    />
    <Component
      className="col-md-4"
      label="Requested Date of Service"
      name="detail-1.serviceDate"
    />
    <Component
      className="col-md-4"
      label="Quantity"
      name="detail-1.quantity"
    />
    <Component
      className="col-md-4"
      label="Monthly Recurring Cost (MRC)"
      name="detail-1.recurringCost"
    />
    <Component
      className="col-md-4"
      label="Non Recurring Cost (NRC)"
      name="detail-1.nonRecurringCost"
    />
    <Component
      className="col-md-4"
      label="State Contract Number"
      name="detail-1.stateContractNumber"
    />
    <Component
      className="col-md-4"
      label="Feature ID / USOC / Product ID"
      name="detail-1.featureId"
    />
    <Component
      className="col-md-4"
      label="Exisiting Billing Account Number"
      name="detail-1.billingAccount"
    />
    <Component
      className="col-md-12"
      label="Description"
      name="detail-1.orderDescription"
      tag="textarea"
    />
    <Component
      className="col-md-12"
      label="Comment"
      name="detail-1.orderComment"
      tag="textarea"
    />
    <div
      className="col-md-12"
    >
      <Button
        active={false}
        className="float-right d-print-none"
        iconProps={
          Object {
            "name": "close-line",
          }
        }
        onClick={[Function]}
        tag="button"
        text="Remove Request #1"
        type="button"
        variant="default"
      />
    </div>
    <div
      className="col-md-12"
    >
      <hr />
    </div>
  </div>
  <div
    aria-labelledby="section7__header--1"
    className="d-flex flex-wrap w-100"
    data-testid="detail-1"
    key="detail-2"
    role="group"
  >
    <div
      className="col-md-12"
      id="section7__header--1"
    >
      Request #2
    </div>
    <Component
      className="col-md-8"
      inline={true}
      label="Order Type"
      name="detail-2.orderType"
      options={
        Array [
          Object {
            "label": "Add",
            "value": "add",
          },
          Object {
            "label": "Disconnect",
            "value": "disconnect",
          },
          Object {
            "label": "Change",
            "value": "change",
          },
          Object {
            "label": "Move",
            "value": "move",
          },
        ]
      }
    />
    <Component
      className="col-md-4"
      label="Requested Date of Service"
      name="detail-2.serviceDate"
    />
    <Component
      className="col-md-4"
      label="Quantity"
      name="detail-2.quantity"
    />
    <Component
      className="col-md-4"
      label="Monthly Recurring Cost (MRC)"
      name="detail-2.recurringCost"
    />
    <Component
      className="col-md-4"
      label="Non Recurring Cost (NRC)"
      name="detail-2.nonRecurringCost"
    />
    <Component
      className="col-md-4"
      label="State Contract Number"
      name="detail-2.stateContractNumber"
    />
    <Component
      className="col-md-4"
      label="Feature ID / USOC / Product ID"
      name="detail-2.featureId"
    />
    <Component
      className="col-md-4"
      label="Exisiting Billing Account Number"
      name="detail-2.billingAccount"
    />
    <Component
      className="col-md-12"
      label="Description"
      name="detail-2.orderDescription"
      tag="textarea"
    />
    <Component
      className="col-md-12"
      label="Comment"
      name="detail-2.orderComment"
      tag="textarea"
    />
    <div
      className="col-md-12"
    >
      <Button
        active={false}
        className="float-right d-print-none"
        iconProps={
          Object {
            "name": "close-line",
          }
        }
        onClick={[Function]}
        tag="button"
        text="Remove Request #2"
        type="button"
        variant="default"
      />
    </div>
  </div>
</div>
`;

exports[`Section7 should render correctly 1`] = `
<FieldArray
  component={[Function]}
  name="orderDetails"
/>
`;
